PSReadLine/ReadLine.vi.cs

```diff
@@ -611,8 +611,19 @@ public static void ViInsertAtEnd(ConsoleKeyInfo? key = null, object arg = null)
         /// </summary>
         public static void ViInsertWithAppend(ConsoleKeyInfo? key = null, object arg = null)
         {
+            _singleton._groupUndoHelper.StartGroup(ViInsertWithAppend, arg);

```

Comment:
This is a public exposed method that can be accociated with keybindings. If we start an editing group here, where will we call `EndGroup`?

I don't quite understand why `StartGroup` is added here. You mentioned in the PR descirption that:
> When authoring this PR I introduced a regression that boils down to the function ViInsertWithAppend.
**This method does not start an undo group, and therefore, when undoing, only the last character appended is removed from the buffer.**

Is the bold part an issue today without your changes? or is it as a result of the changes in this PR?

Reply:
In fact, the first commit introduces a regression. Closing a group if not finding a matching character in <kbd>c</kbd><kbd>f</kbd><kbd>?</kbd> _is_ the regression.

Indeed, this sequence is followed by a call to `ViInsertWithAppend` but relies on the group started in `ViReplaceToChar`.

On the one hand, not closing the group makes PSRL crash if the same sequence is retried (because nested groups are not supported). On the other hand, closing the group prevents _appends_ from recording the whole future sequence of inserts to undo.

So I'm not sure how best to solve this problem.

Perhaps we should not throw an exception when a group is started without closing the previous one. Just carrying along with recording sequences to undo ?

Reply:
@daxian-dbw also, I mostly did the change out of symmetry with the method `ViInsertWithDelete` which, itself, starts a group. So I think the same argument should apply.

If this is a breaking change, I’m convinced it is a _necessary_ change.
What do you think ?

Reply:
Sorry for the looooonnnnggg delay in reviewing this PR :(
While re-reading the code changes, I think maybe the best way to solve this problem is to properly support nested groups. It's a way bigger change I guess but I just wonder maybe that's the right way forward if we want to eliminate all issues related to unbalanced edit groups ...

Reply:
OK, I started from scratch and ended up supporting multiple starts for one single end.
I think there can only be one edit group at any single time. It’s just that methods may be called in multiple orders.

This allows to support adding a start in the public facing method without impact.

I fear the that undo support is a bit brittle as it now needs a ton of states to track.
Please, let me know what you feel.

Reply:
The new change supports closing all the "starts" so that would no longer be an issue.